packages/cli/generation/ir-generator/src/utils/getCyclicTypes.ts

```diff
@@ -0,0 +1,203 @@
+import { ResolvedType } from "../resolvers/ResolvedType";
+import { TypeResolver } from "../resolvers/TypeResolver";
+import {
+    CASINGS_GENERATOR,
+    getAllPropertiesForType,
+    ObjectPropertyWithPath,
+    TypeName,
+    TypePropertyArgs
+} from "./getAllPropertiesForObject";
+
+export declare type CyclicTypeArgs = TypePropertyArgs & {
+    childName: TypeName;
+    seen?: Set<TypeName>;
+};
+
+declare type TypeOrProperty = ResolvedTypeWithKeys | ObjectPropertyWithPath;
+
+declare type UnionValue =
+    | string
+    | {
+          docs?: string | undefined;
+          name?: string | undefined;
+          type?: string | Record<string, never> | undefined;
+          key?: string | { value: string; name?: string | undefined } | undefined;
+      };
+
+interface ResolvedTypeWithKeys {
+    keys: string[];
+    type: ResolvedType | null;
+}
+
+interface TypeParams {
+    keys: string[];
+    name: TypeName;
+    type: ResolvedType;
+    filePath: RelativeFilePath;
+}
+
+/**
+ * This function will return the first found path to a named child type from a given type. It will only search within the file of the given type.
+ * @param args - The arguments to the function
+ * @returns String array of keys and objects to the named child type, or null if not found
+ */
+export function getCyclicTypes(args: CyclicTypeArgs): string[] | null {
+    args.seen ??= new Set<TypeName>([args.typeName]);
+    const fileContext: FernFileContext = constructFernFileContext({
+        relativeFilepath: args.filepathOfDeclaration,
+        definitionFile: args.definitionFile,
+        rootApiFile: args.workspace.definition.rootApiFile.contents,
+        casingsGenerator: CASINGS_GENERATOR
+    });
+    const resolvedType: ResolvedType | undefined = args.typeResolver.resolveType({
+        type: args.typeName,
+        file: fileContext
+    });
+    if (resolvedType == null) {
+        return null;
+    }
+    if (resolvedType._type !== "named") {
+        return null;
+    }
+
+    // If this is a union type, every object path must be cyclic in order for the type to be considered an invalid cyclic type. This is also why we ignore container types, since they can be empty.
+    const isUnion =
+        isRawDiscriminatedUnionDefinition(resolvedType.declaration) ||
+        isRawUndiscriminatedUnionDefinition(resolvedType.declaration);
+
+    const typeProperties: TypeOrProperty[] = isUnion
+        ? unboxUnionTypes(resolvedType, fileContext, args.typeResolver)
+        : unboxObjectType(args, resolvedType);
+
+    return isUnion
+        ? checkUnionChildTypes(args, typeProperties.flatMap(getTypeParams))
+        : checkObjectChildTypes(args, typeProperties.flatMap(getTypeParams));
+}
+
+function unboxObjectType(args: CyclicTypeArgs, t: ResolvedType.Named): TypeOrProperty[] {
+    if (isRawObjectDefinition(t.declaration)) {
+        return getAllPropertiesForType(args) as TypeOrProperty[];
```

Comment:
why do we have to use a cast?

Reply:
Not really necessary, but the linter on my editor likes casting to the union object rather than returning an array of elements which are either the type object or the properties object. I think it makes it fairly easy to parse, but it should run fine without it if we want to remove the cast here.

Reply:
took it out in the latest revision, should be all good now